File: routedev/desktop/shared/remote-protocol.ts
// RouteDev Remote v1 协议类型。
// 本文件只包含可序列化类型与稳定常量，不依赖 Electron、Node HTTP 或业务实现。

export const REMOTE_PROTOCOL_VERSION = 1 as const;
export const REMOTE_DEFAULT_PORT = 43117 as const;

export const REMOTE_API_ROUTES = {
  health: { method: 'GET', path: '/v1/health' },
  capabilities: { method: 'GET', path: '/v1/capabilities' },
  pair: { method: 'POST', path: '/v1/pair' },
  devices: { method: 'GET', path: '/v1/devices' },
  revokeDevice: { method: 'DELETE', path: '/v1/devices/{deviceId}' },
  sessions: { method: 'GET', path: '/v1/sessions' },
  createSession: { method: 'POST', path: '/v1/sessions' },
  session: { method: 'GET', path: '/v1/sessions/{sessionId}' },
  timeline: { method: 'GET', path: '/v1/sessions/{sessionId}/timeline' },
  sendMessage: { method: 'POST', path: '/v1/sessions/{sessionId}/messages' },
  stopTask: { method: 'POST', path: '/v1/sessions/{sessionId}/stop' },
  resolveApproval: { method: 'POST', path: '/v1/approvals/{approvalId}/resolve' },
  skills: { method: 'GET', path: '/v1/skills' },
  mcpServers: { method: 'GET', path: '/v1/mcp/servers' },
  tools: { method: 'GET', path: '/v1/tools' },
  events: { method: 'GET', path: '/v1/events' },
} as const;

export const REMOTE_ERROR_CODES = [
  'AUTH_REQUIRED',
  'AUTH_REVOKED',
  'PAIRING_EXPIRED',
  'PAIRING_INVALID',
  'SCOPE_DENIED',
  'SESSION_NOT_FOUND',
  'SESSION_BUSY',
  'SKILL_NOT_AVAILABLE',
  'MCP_NOT_AVAILABLE',
  'TOOL_NOT_ALLOWED',
  'APPROVAL_EXPIRED',
  'CONFLICT',
  'RATE_LIMITED',
  'ENGINE_UNAVAILABLE',
  'PROTOCOL_MISMATCH',
] as const;

export type RemoteErrorCode = typeof REMOTE_ERROR_CODES[number];

export const REMOTE_DEVICE_SCOPES = [
  'sessions:read',
  'messages:send',
  'tasks:stop',
  'approvals:resolve',
  'skills:select',
  'mcp:select',
  'autonomy:change',
] as const;

export type RemoteDeviceScope = typeof REMOTE_DEVICE_SCOPES[number];

export const REMOTE_EVENT_TYPES = [
  'session.created',
  'session.updated',
  'turn.started',
  'assistant.text.delta',
  'assistant.reasoning.delta',
  'assistant.progress',
  'tool.started',
  'tool.output.delta',
  'tool.completed',
  'tool.failed',
  'todo.replaced',
  'todo.updated',
  'approval.required',
  'approval.resolved',
  'turn.completed',
  'turn.failed',
  'task.completed',
  'task.failed',
  'connection.notice',
] as const;

export type RemoteEventType = typeof REMOTE_EVENT_TYPES[number];
export type RemoteAutonomyMode = 'auto' | 'semi' | 'manual';
export type RemoteSessionStatus = 'idle' | 'running' | 'waiting_approval' | 'completed' | 'failed';
export type RemoteTodoStatus = 'pending' | 'in_progress' | 'completed';

export interface RemoteError {
  code: RemoteErrorCode;
  message: string;
  retryable: boolean;
  details?: Record<string, unknown>;
}

export interface RemoteApiResponse<T> {
  protocolVersion: typeof REMOTE_PROTOCOL_VERSION;
  requestId: string;
  timestamp: string;
  sessionId: string | null;
  turnId: string | null;
  ok: boolean;
  payload: T | null;
  error: RemoteError | null;
}

export interface RemoteEventEnvelope<TType extends RemoteEventType = RemoteEventType, TPayload = unknown> {
  protocolVersion: typeof REMOTE_PROTOCOL_VERSION;
  eventId: string;
  timestamp: string;
  sessionId: string;
  turnId: string | null;
  sequence: number;
  type: TType;
  payload: TPayload;
}

export interface RemoteHealth {
  status: 'ok' | 'degraded';
  desktopName: string;
  gatewayVersion: string;
  engineAvailable: boolean;
}

export interface RemoteCapabilities {
  protocolVersion: typeof REMOTE_PROTOCOL_VERSION;
  eventTypes: RemoteEventType[];
  deviceScopes: RemoteDeviceScope[];
  supportsImages: boolean;
  maxImageCount: number;
  maxImageBytes: number;
}

export interface RemotePairRequest {
  pairingId: string;
  secret: string;
  deviceId: string;
  deviceName: string;
  requestedScopes: RemoteDeviceScope[];
  protocolVersion: typeof REMOTE_PROTOCOL_VERSION;
}

export interface RemotePairResponse {
  deviceId: string;
  deviceToken: string;
  grantedScopes: RemoteDeviceScope[];
  desktopName: string;
  baseUrl: string;
}

export interface RemoteDevice {
  deviceId: string;
  deviceName: string;
  scopes: RemoteDeviceScope[];
  createdAt: string;
  lastSeenAt: string | null;
  revokedAt: string | null;
}

export interface RemoteSessionSummary {
  sessionId: string;
  title: string;
  status: RemoteSessionStatus;
  createdAt: string;
  updatedAt: string;
  activeTurnId: string | null;
  lastSequence: number;
}

export interface RemoteSessionDetail extends RemoteSessionSummary {
  projectId: string | null;
  projectName: string | null;
  latestResult: string | null;
}

export interface RemoteCreateSessionRequest {
  title?: string;
  projectId?: string;
  clientSessionId: string;
}

export interface RemoteCreateSessionResponse {
  session: RemoteSessionDetail;
  clientSessionId: string;
}

export interface RemoteImageInput {
  mediaType: 'image/png' | 'image/jpeg' | 'image/webp';
  dataBase64: string;
  filename?: string;
}

export interface RemoteSendMessageRequest {
  text: string;
  images?: RemoteImageInput[];
  skillIds?: string[];
  mcpServerIds?: string[];
  allowedToolNames?: string[];
  autonomyMode?: RemoteAutonomyMode;
  clientMessageId: string;
}

export interface RemoteSendMessageResponse {
  sessionId: string;
  turnId: string;
  clientMessageId: string;
  acceptedAt: string;
  duplicate: boolean;
}

export interface RemoteStopTaskRequest {
  turnId?: string;
  reason?: string;
}

export interface RemoteApprovalResolveRequest {
  approved: boolean;
  payload?: unknown;
}

export interface RemoteSkill {
  id: string;
  name: string;
  description: string;
  source: string;
  enabled: boolean;
}

export interface RemoteMcpServer {
  id: string;
  name: string;
  connected: boolean;
  toolCount: number;
}

export interface RemoteTool {
  name: string;
  description: string;
  source: 'builtin' | 'mcp';
  mcpServerId: string | null;
  allowed: boolean;
}

export interface RemoteTodoItem {
  id: string;
  content: string;
  status: RemoteTodoStatus;
  priority?: 'low' | 'medium' | 'high';
}

export interface RemoteApproval {
  approvalId: string;
  toolName: string;
  summary: string;
  expiresAt: string;
  canResolveRemotely: boolean;
}

export interface RemoteEventPayloadMap {
  'session.created': { session: RemoteSessionSummary };
  'session.updated': { session: RemoteSessionSummary };
  'turn.started': { clientMessageId: string; userText: string };
  'assistant.text.delta': { text: string };
  'assistant.reasoning.delta': { text: string; visibility: 'provider_returned' };
  'assistant.progress': { text: string; source: 'model' | 'engine' };
  'tool.started': { toolCallId: string; toolName: string; argsSummary: string | null };
  'tool.output.delta': { toolCallId: string; delta: string };
  'tool.completed': { toolCallId: string; outputSummary: string | null };
  'tool.failed': { toolCallId: string; error: RemoteError };
  'todo.replaced': { items: RemoteTodoItem[] };
  'todo.updated': { item: RemoteTodoItem };
  'approval.required': { approval: RemoteApproval };
  'approval.resolved': { approvalId: string; approved: boolean; resolvedBy: 'desktop' | 'android' };
  'turn.completed': { finishReason: string; completionStatus: string | null };
  'turn.failed': { error: RemoteError };
  'task.completed': { summary: string };
  'task.failed': { error: RemoteError };
  'connection.notice': { level: 'info' | 'warning' | 'error'; message: string; code: string | null };
}

export type AnyRemoteEvent = {
  [K in RemoteEventType]: RemoteEventEnvelope<K, RemoteEventPayloadMap[K]>;
}[RemoteEventType];

export interface RemoteTimelineResponse {
  events: AnyRemoteEvent[];
  nextSequence: number;
}

/**
 * 校验已经反序列化的事件是否能组成严格有序、无重复的单会话时间线。
 * 网络层和本地数据库都使用相同语义：sequence 是唯一排序依据。
 */
export function validateRemoteEventSequence(events: readonly RemoteEventEnvelope[]): string[] {
  const errors: string[] = [];
  const eventIds = new Set<string>();
  let previousSequence = 0;
  let sessionId: string | null = null;

  for (const event of events) {
    if (eventIds.has(event.eventId)) {
      errors.push(`duplicate eventId: ${event.eventId}`);
    }
    eventIds.add(event.eventId);

    if (sessionId === null) sessionId = event.sessionId;
    if (event.sessionId !== sessionId) {
      errors.push(`mixed sessionId: ${event.sessionId}`);
    }
    if (event.sequence <= previousSequence) {
      errors.push(`sequence must increase: ${previousSequence} -> ${event.sequence}`);
    }
    previousSequence = event.sequence;
  }
  return errors;
}

File: routedev-android/protocol/src/main/kotlin/com/routedev/remote/protocol/RemoteProtocol.kt
package com.routedev.remote.protocol

import java.time.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

const val REMOTE_PROTOCOL_VERSION: Int = 1

@Serializable
enum class RemoteErrorCode {
    AUTH_REQUIRED,
    AUTH_REVOKED,
    PAIRING_EXPIRED,
    PAIRING_INVALID,
    SCOPE_DENIED,
    SESSION_NOT_FOUND,
    SESSION_BUSY,
    SKILL_NOT_AVAILABLE,
    MCP_NOT_AVAILABLE,
    TOOL_NOT_ALLOWED,
    APPROVAL_EXPIRED,
    CONFLICT,
    RATE_LIMITED,
    ENGINE_UNAVAILABLE,
    PROTOCOL_MISMATCH,
}

@Serializable
enum class RemoteDeviceScope {
    @SerialName("sessions:read") SESSIONS_READ,
    @SerialName("messages:send") MESSAGES_SEND,
    @SerialName("tasks:stop") TASKS_STOP,
    @SerialName("approvals:resolve") APPROVALS_RESOLVE,
    @SerialName("skills:select") SKILLS_SELECT,
    @SerialName("mcp:select") MCP_SELECT,
    @SerialName("autonomy:change") AUTONOMY_CHANGE,
}

@Serializable
enum class RemoteEventType {
    @SerialName("session.created") SESSION_CREATED,
    @SerialName("session.updated") SESSION_UPDATED,
    @SerialName("turn.queued") TURN_QUEUED,
    @SerialName("turn.started") TURN_STARTED,
    @SerialName("assistant.text.delta") ASSISTANT_TEXT_DELTA,
    @SerialName("assistant.reasoning.delta") ASSISTANT_REASONING_DELTA,
    @SerialName("assistant.progress") ASSISTANT_PROGRESS,
    @SerialName("tool.started") TOOL_STARTED,
    @SerialName("tool.output.delta") TOOL_OUTPUT_DELTA,
    @SerialName("tool.completed") TOOL_COMPLETED,
    @SerialName("tool.failed") TOOL_FAILED,
    @SerialName("todo.replaced") TODO_REPLACED,
    @SerialName("todo.updated") TODO_UPDATED,
    @SerialName("approval.required") APPROVAL_REQUIRED,
    @SerialName("approval.resolved") APPROVAL_RESOLVED,
    @SerialName("turn.completed") TURN_COMPLETED,
    @SerialName("turn.failed") TURN_FAILED,
    @SerialName("task.completed") TASK_COMPLETED,
    @SerialName("task.failed") TASK_FAILED,
    @SerialName("connection.notice") CONNECTION_NOTICE,
}

@Serializable
enum class RemoteAutonomyMode {
    @SerialName("auto") AUTO,
    @SerialName("semi") SEMI,
    @SerialName("manual") MANUAL,
}

@Serializable
data class RemoteError(
    val code: RemoteErrorCode,
    val message: String,
    val retryable: Boolean,
    val details: JsonObject? = null,
)

@Serializable
data class RemoteApiResponse(
    val protocolVersion: Int,
    val requestId: String,
    val timestamp: String,
    val sessionId: String?,
    val turnId: String?,
    val ok: Boolean,
    val payload: JsonElement?,
    val error: RemoteError?,
)

@Serializable
data class RemoteEventEnvelope(
    val protocolVersion: Int,
    val eventId: String,
    val timestamp: String,
    val sessionId: String,
    val turnId: String?,
    val sequence: Long,
    val type: RemoteEventType,
    val payload: JsonObject,
)

@Serializable
data class RemotePairRequest(
    val pairingId: String,
    val secret: String,
    val deviceId: String,
    val deviceName: String,
    val requestedScopes: List<RemoteDeviceScope>,
    val protocolVersion: Int,
)

@Serializable
data class RemoteCreateSessionRequest(
    val title: String? = null,
    val projectId: String? = null,
    val clientSessionId: String,
)

@Serializable
data class RemoteImageInput(
    val mediaType: String,
    val dataBase64: String,
    val filename: String? = null,
)

@Serializable
data class RemoteSendMessageRequest(
    val text: String,
    val images: List<RemoteImageInput>? = null,
    val skillIds: List<String>? = null,
    val mcpServerIds: List<String>? = null,
    val allowedToolNames: List<String>? = null,
    val autonomyMode: RemoteAutonomyMode? = null,
    val clientMessageId: String,
)

@Serializable
data class RemoteStopTaskRequest(
    val turnId: String? = null,
    val reason: String? = null,
)

@Serializable
data class RemoteApprovalResolveRequest(
    val approved: Boolean,
    val payload: JsonElement? = null,
)

object RemoteProtocolCodec {
    val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = true
        encodeDefaults = true
    }

    fun decodeEvent(text: String): RemoteEventEnvelope =
        json.decodeFromString<RemoteEventEnvelope>(text).also(::validateEvent)

    fun decodeResponse(text: String): RemoteApiResponse =
        json.decodeFromString<RemoteApiResponse>(text).also(::validateResponse)

    fun validateSequence(events: List<RemoteEventEnvelope>): List<String> {
        val errors = mutableListOf<String>()
        val eventIds = mutableSetOf<String>()
        var previous = 0L
        var sessionId: String? = null

        events.forEach { event ->
            if (!eventIds.add(event.eventId)) errors += "duplicate eventId: ${event.eventId}"
            if (sessionId == null) sessionId = event.sessionId
            if (sessionId != event.sessionId) errors += "mixed sessionId: ${event.sessionId}"
            if (event.sequence <= previous) {
                errors += "sequence must increase: $previous -> ${event.sequence}"
            }
            previous = event.sequence
        }
        return errors
    }

    private fun validateEvent(event: RemoteEventEnvelope) {
        require(event.protocolVersion == REMOTE_PROTOCOL_VERSION) { "protocol mismatch" }
        require(event.eventId.isNotBlank()) { "eventId is required" }
        require(event.sessionId.isNotBlank()) { "sessionId is required" }
        require(event.sequence > 0) { "sequence must be positive" }
        Instant.parse(event.timestamp)
    }

    private fun validateResponse(response: RemoteApiResponse) {
        require(response.protocolVersion == REMOTE_PROTOCOL_VERSION) { "protocol mismatch" }
        require(response.requestId.isNotBlank()) { "requestId is required" }
        require(response.ok == (response.error == null)) { "ok/error fields conflict" }
        Instant.parse(response.timestamp)
    }
}
